content/the-basics-of-go.md

```diff
@@ -0,0 +1,865 @@
+}
+func main() {
+  go count(0)
+  go count(8)
+  // We need something to hold the main thread,
+  // otherwise the program will exit immediately
+  fmt.Scanln()
+}
+```
+
+The output of the code above wouldn't be exacly 0 to 4, then 8 to 13. They might overlap each other in a way
+that this output would occur.
+
+```
+0
+1
+8
+9
+2
+3
+4
+10
+11
+12
+```
+
+In the real world, goroutines are used to do multiple things at the same time, background work, or anything
+that would require very fast stuffs. A simple example (well at least, for me) is a captcha bot. If you visit
+the Teknologi Umum's Telegram group, the first thing that you'd meet is the captcha bot. It does basic
+captcha validation for each joined user.
+
+But the question now is, how do you handle that if there are multiple users joining at the same time? On a normal
+code, the captcha bot would just wait for each user to solve the captcha, then move to the next user, right?
+But is it ideal? What if the second user that's joined just gave a spam message? It would be a total chaos.
+
+So, I came up with this flow:
+
+1. A user joined the group. The captcha bot will send a captcha to the user.
+2. The captcha bot calls a function that start a 60 seconds timer for the user to complete the captcha with goroutine.
+3. If the 60 seconds has passed, and the user haven't solved the captcha, the captcha bot will kick the user.
+4. Listen to any incoming message by anyone, and if the message was sent by one of the user that need to solve the captcha,
+   we will validate their input.
+5. If the user solved the captcha, we will send a message to the user that they have solved the captcha. And remove
+   the user's captcha from the list of active captcha.
+
+It's that simple, but effective. An additional use for the goroutine is to automatically delete the message that was sent
+by both the user and the bot. The code of the system above is available on [Github](https://github.com/teknologi-umum/captcha/tree/master/captcha).
+
+## Channel
+
+Let's dive about one more thing before we end this article: Channel.
+
+Channel is a way to communicate between goroutines. To use it, simply make a channel of a type, and then
+use the channel as usual. The channel is a type that can be used to send and receive values. Let's see the example
+code so you won't have to bother with all the technical terms.
+
+```go
+package main
+
+func giveValue(c chan string, v string) {
+  // This is how you would send a value
+  // into a channel.
+  //
+  // c is an argument variable (see function signature)
+  // that holds a channel of string.
+  // v is also an argument variable, that's
+  // a type of string.
+  c <- v
+}
+
+func main() {
+  c := make(chan string)
+
+  go giveValue(c, "hello")
+  go giveValue(c, "world")
+
+  // <-c means we receive a value from the channel c
+  // into the result variable.
+  //
+  // Well, you can also do directly fmt.Println(<-c)
+  // but that would confuse you.
+  result := <-c
```

Comment:
Should probably tell that `<-c` is blocking until it receives a value.

Reply:
yeah sure